File: app/src/main/java/com/mohsenoid/rickandmorty/ui/util/ImageDownloaderImpl.java
package com.mohsenoid.rickandmorty.ui.util;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;

import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.mohsenoid.rickandmorty.data.service.network.NetworkHelper;
import com.mohsenoid.rickandmorty.executor.TaskExecutor;

import java.io.File;
import java.io.IOException;

public class ImageDownloaderImpl implements ImageDownloader {

    private static final String SEPARATOR = "/";

    @VisibleForTesting
    public static ImageDownloaderImpl instance;

    private final NetworkHelper networkHelper;
    private final String cacheDirectoryPath;
    private final TaskExecutor ioTaskExecutor;
    private final TaskExecutor mainTaskExecutor;

    private ImageDownloaderImpl(NetworkHelper networkHelper, String cacheDirectoryPath, TaskExecutor ioTaskExecutor, TaskExecutor mainTaskExecutor) {
        this.networkHelper = networkHelper;
        this.cacheDirectoryPath = cacheDirectoryPath;
        this.ioTaskExecutor = ioTaskExecutor;
        this.mainTaskExecutor = mainTaskExecutor;
    }

    public static synchronized ImageDownloaderImpl getInstance(NetworkHelper networkHelper, String cacheDirectoryPath, TaskExecutor ioTaskExecutor, TaskExecutor mainTaskExecutor) {
        if (instance == null)
            instance = new ImageDownloaderImpl(networkHelper, cacheDirectoryPath, ioTaskExecutor, mainTaskExecutor);

        return instance;
    }

    @Override
    public void downloadImage(String imageUrl, ImageView imageView, @Nullable ProgressBar progress) {
        imageView.setVisibility(View.INVISIBLE);
        if (progress != null) progress.setVisibility(View.VISIBLE);

        ioTaskExecutor.execute(() -> {
            String fileName = extractFileName(imageUrl);
            File imageFile = new File(cacheDirectoryPath + File.separator + fileName);


            if (!imageFile.exists()) {
                try {
                    networkHelper.requestImageData(imageUrl, imageFile);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            Bitmap bitmap = loadBitmapFile(imageFile);

            mainTaskExecutor.execute(() -> {
                if (progress != null) progress.setVisibility(View.GONE);
                imageView.setVisibility(View.VISIBLE);
                imageView.setImageBitmap(bitmap);
            });
        });
    }

    @Override
    public String extractFileName(String url) {
        if (url == null) return null;

        String[] urlParts = url.split(SEPARATOR);
        return urlParts[urlParts.length - 1];
    }

    private Bitmap loadBitmapFile(File imageFile) {
        try {
            return BitmapFactory.decodeFile(imageFile.getAbsolutePath());
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/util/image/ImageDownloaderImpl.kt
package com.mohsenoid.rickandmorty.util.image

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import androidx.core.view.isVisible
import com.mohsenoid.rickandmorty.data.network.NetworkHelper
import com.mohsenoid.rickandmorty.util.dispatcher.DispatcherProvider
import kotlinx.coroutines.withContext
import timber.log.Timber
import java.io.File
import java.io.IOException

class ImageDownloaderImpl(
    private val networkHelper: NetworkHelper,
    private val cacheDirectoryPath: String,
    private val dispatcherProvider: DispatcherProvider
) : ImageDownloader {

    override suspend fun downloadImage(
        imageUrl: String,
        imageView: ImageView,
        progress: ProgressBar?
    ) {
        imageView.visibility = View.INVISIBLE
        progress?.isVisible = true

        withContext(dispatcherProvider.ioDispatcher) {
            val fileName = extractFileName(imageUrl)
            val imageFile = File(cacheDirectoryPath + File.separator + fileName)
            if (!imageFile.exists()) {
                try {
                    networkHelper.requestImageData(imageUrl, imageFile)
                } catch (e: IOException) {
                    e.printStackTrace()
                    if (imageFile.delete()) Timber.w("Image file ${imageFile.name} deleted!")
                }
            }
            val bitmap = loadBitmapFile(imageFile, imageView.width, imageView.height)
            withContext(dispatcherProvider.mainDispatcher) {
                progress?.isVisible = false
                imageView.isVisible = true
                imageView.setImageBitmap(bitmap)
            }
        }
    }

    override fun extractFileName(url: String): String {
        return url.split(SEPARATOR).last()
    }

    private fun loadBitmapFile(imageFile: File, outWidth: Int, outHeight: Int): Bitmap? {
        return try {
            val options = BitmapFactory.Options()
            options.outWidth = outWidth
            options.outHeight = outHeight
            BitmapFactory.decodeFile(imageFile.absolutePath, options)
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    companion object {
        private const val SEPARATOR = "/"
    }
}
